src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -113,6 +123,28 @@ public bool HasValue
             get { return Buffer != null; }
         }
 
+        /// <summary>
+        /// Gets the <see cref="char"/> at a specified position in the current <see cref="StringSegment"/>.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <returns></returns>
+        public char this[int index]
+        {
+            get
+            {
+                if (!HasValue)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+                if (index - Offset >= Length)
```

Comment:
This should match the offset checks in the ctor:

https://github.com/aspnet/Common/blob/b84883966fccb1b7b086f93385680405a6b66aad/src/Microsoft.Extensions.Primitives/StringSegment.cs#L46